### PR TITLE
Hyperparameter search to train transition

### PR DESCRIPTION
**Pull request by Jakoviz**

### CODE REVIEW

File: hyperparameter_search.py [hunk 1]
```diff
@@ -297,17 +298,17 @@ def objective(trial, MODEL_NAME, NUM_CLASSES, N_EPOCHS, OPTIMIZER_SEARCH_SPACE,
...
+@click.option('-e', '--no_of_epochs', type=int, show_default=True, default=20, help='Number of epochs in training loop.')
+@click.option('-es', '--early_stopping_counter', type=int, help='Number of consequtive epochs with no improvement in loss until trial is stopped. Default: ~one third of no of epochs.')
+@click.option('-t', '--no_of_trials', type=int, show_default=True, default=20, help='Number of hyperparamter search trials in training loop.')
 @click.option('-d', '--dataset', type=click.Choice(['plant', 'plant_golden', 'leaf'], case_sensitive=False), default="plant", help='Already available dataset to use to train the model. Give either -d or -csv, not both.')
 @click.option('-csv', '--data-csv', type=str, help='Full file path to dataset CSV-file created during segmentation. Give either -d or -csv, not both.')
 @click.option('-b', '--binary', is_flag=True, show_default=True, default=False, help='Train binary classifier instead of multiclass classifier.')
 @click.option('-aug', '--augmentation', is_flag=True, show_default=True, default=True, help='Use data-augmentation for the training.')
```
Comment (Redande) [inline hyperparameter_search.py:307]:
```suggestion
@click.option('-aug/-no-aug', '--augmentation/--no-augmentation', show_default=True, default=True, help='Use data-augmentation for the training.')
```

Comment (Redande) [reply hyperparameter_search.py:307]: I think this currently has the same issue @perander brought up for train.py, which is that by default it is on and adding the actual `-aug` flag sets it to False. Suggested change is how @jarsba fixed it for train.py

File: hyperparameter_search.py [hunk 2]
```diff
@@ -394,19 +403,29 @@ def search_hyperparameters(model, no_of_epochs, no_of_trials, dataset, data_csv,
...
-    ## which can then be used in train.py to evaluate the model.
-    # study.trials_dataframe()
-    # study.best_trial
-    # study.best_params
-    # study.best_trials
-    # torch.save(test_dataset, f"{MODEL_NAME}_test_dataset_with_train+valid_best_value_{study.best_value}.pt")
-    # torch.save(model.state_dict(), f'{MODEL_NAME}_weights_with_train+valid_best_value_{study.best_value}.pt')
+
+    df = study.trials_dataframe()
+    df = df.sort_values(by=['value'], ascending=True).iloc[0:9,:]
```
Comment (Redande) [inline hyperparameter_search.py:419]: This currently selects the 10 rows with the smallest value in the `value`-column. I'm not sure what values are in the column, just doublechecking that we want to select the 10 smallest and not the 10 largest (which could be achieved by changing to `ascending=False`)

File: train.py [hunk 3]
```diff
@@ -42,13 +43,22 @@
...
+@click.option('-aug', '--augmentation', is_flag=True, show_default=True, default=True, help='Use data-augmentation for the training.')
+@click.option('-s', '--save', is_flag=True, show_default=True, default=True, help='Save the trained model and add information to model dataframe.')
+=======
 @click.option('-bl', '--binary-label', type=int, help='Binary label when dataset has more than two labels. Classification is done using one-vs-rest, where the binary label corresponds to the one compared to other labels.')
 @click.option('-p', '--params-file', type=str, default="hyperparams.yaml", help='Full file path to hyperparameter-file used during the training. File must be a YAMl file and similarly structured than hyperparams.yaml.')
 @click.option('-aug/-no-aug', '--augmentation/--no-augmentation', show_default=True, default=True, help='Use data-augmentation for the training.')
 @click.option('-s/-nos', '--save/--no-save', show_default=True, default=True, help='Save the trained model and add information to model dataframe.')
+>>>>>>> cb0716d121897bb2463d55ac10218bd0c717783f
```
Comment (Redande) [inline train.py:55]:
```suggestion
@click.option('-bl', '--binary-label', type=int, help='Binary label when dataset has more than two labels. Classification is done using one-vs-rest, where the binary label corresponds to the one compared to other labels.')
@click.option('-p', '--params-file', type=str, default="hyperparams.yaml", help='Full file path to hyperparameter-file used during the training. File must be a YAMl file and similarly structured than hyperparams.yaml.')
@click.option('-aug/-no-aug', '--augmentation/--no-augmentation', show_default=True, default=True, help='Use data-augmentation for the training.')
@click.option('-s/-nos', '--save/--no-save', show_default=True, default=True, help='Save the trained model and add information to model dataframe.')
```

Comment (Redande) [reply train.py:55]: Unresolved merge conflict. The second of the options is what we have in main-branch.

File: train.py [hunk 4]
```diff
@@ -95,6 +105,14 @@ def train(model, dataset, data_csv, binary, binary_label, params_file, augmentat
...
             raise exc
 
+
+    # hyperparameters:
+    N_EPOCHS = int(params[MODEL_NAME]['N_EPOCHS'])
+    BATCH_SIZE_TRAIN = int(params[MODEL_NAME]['BATCH_SIZE_TRAIN'])
+    BATCH_SIZE_TEST = int(params[MODEL_NAME]['BATCH_SIZE_TEST'])
+    OPTIMIZER = params[MODEL_NAME]['OPTIMIZER']
+    LR = float(params[MODEL_NAME]['LR'])
+    WEIGHT_DECAY = float(params[MODEL_NAME]['WEIGHT_DECAY'])
```
Comment (Redande) [inline train.py:115]:
```suggestion
```

Comment (Redande) [reply train.py:115]: These come later at lines 211-216 in the else-branch for NN models. BOVW-model doesn't have all of these hyperparameters so they can't be defined here.

File: train.py [hunk 5]
```diff
@@ -124,64 +142,111 @@ def train(model, dataset, data_csv, binary, binary_label, params_file, augmentat
...
+        parameter_grid['dampening'] = float(params[MODEL_NAME]['DAMPENING'])
+        parameter_grid['momentum'] = float(params[MODEL_NAME]['MOMENTUM'])
+        optimizer = optim.SGD(model_class.parameters(), **parameter_grid)
+=======
+    
     if model == 'bag_of_words':
         model_class, y_true, y_pred, test_accuracy, test_loss, other_json = train_bow(master_dataset.df, test_size, NUM_CLASSES, params, save, binary_label)
         train_accuracy = None
         train_loss = None
+>>>>>>> cb0716d121897bb2463d55ac10218bd0c717783f
```
Comment (Redande) [inline train.py:209]:
```suggestion
    if model == 'bag_of_words':
        model_class, y_true, y_pred, test_accuracy, test_loss, other_json = train_bow(master_dataset.df, test_size, NUM_CLASSES, params, save, binary_label)
        train_accuracy = None
        train_loss = None
```

Comment (Redande) [reply train.py:209]: Another unresolved merge conflict. We should keep the latter of the two options, as the stuff in the first option is defined later in the else branch for the NN models, at lines 217-232.

File: train.py [hunk 6]
```diff
@@ -124,64 +142,111 @@ def train(model, dataset, data_csv, binary, binary_label, params_file, augmentat
...
         BATCH_SIZE_TRAIN = int(params[model]['BATCH_SIZE_TRAIN'])
         BATCH_SIZE_TEST = int(params[model]['BATCH_SIZE_TEST'])
         OPTIMIZER = params[model]['OPTIMIZER']
         LR = float(params[model]['LR'])
         WEIGHT_DECAY = float(params[model]['WEIGHT_DECAY'])
-
+        train_dataset, test_dataset = torch.utils.data.random_split(dataset=master_dataset,
+                                        lengths=[train_size + val_size, test_size],
+                                        generator=torch.Generator().manual_seed(42))
         train_dataset, test_dataset = torch.utils.data.random_split(master_dataset, [train_size, test_size])
```
Comment (Redande) [inline train.py:220]:
```suggestion
```

Comment (Redande) [reply train.py:220]: Overrides the previous line where we use the manual seed from hyperparameter search.

File: train.py [hunk 7]
```diff
@@ -124,64 +142,111 @@ def train(model, dataset, data_csv, binary, binary_label, params_file, augmentat
...
+                parameter_grid['momentum'] = float(params[MODEL_NAME]['MOMENTUM'])
+                parameter_grid['alpha'] = float(params[MODEL_NAME]['ALPHA'])
                 optimizer = optim.RMSprop(model_class.parameters(), **parameter_grid)
 
-        loss_function = torch.nn.CrossEntropyLoss()
-
-        # %%
-
-        # training
+          loss_function = torch.nn.CrossEntropyLoss()
```
Comment (Redande) [inline train.py:249]:
```suggestion
      loss_function = torch.nn.CrossEntropyLoss()
```

Comment (Redande) [reply train.py:249]: Incorrect indentation

File: train.py [hunk 8]
```diff
@@ -213,6 +278,16 @@ def train(model, dataset, data_csv, binary, binary_label, params_file, augmentat
...
+                output = model_class(data)
+
+                if len(output) == 2:
+                    output = output.logits
+
+                train_loss = loss_function(output, target)
+                train_loss.backward()
+                optimizer.step()
+
+                pred = output.max(1, keepdim=True)[1]
```
Comment (Redande) [inline train.py:290]:
```suggestion
```

Comment (Redande) [reply train.py:290]: Unnecessary repetition of lines 270-279

File: train.py [hunk 9]
```diff
@@ -223,11 +298,14 @@ def train(model, dataset, data_csv, binary, binary_label, params_file, augmentat
...
 
-
             # Training loss average for all batches
             training_losses.append(total_train_loss / len(train_plant_dataloader))
             training_accuracies.append((100. * train_correct / total))
 
+        ## Training loss average for all batches
+        training_losses.append(total_train_loss / len(train_plant_dataloader))
+        training_accuracies.append((100. * train_correct / total))
+
```
Comment (Redande) [inline train.py:308]:
```suggestion
```

Comment (Redande) [reply train.py:308]: Repetition of lines 301-303, and with incorrect indentation

**Review by Redande — CHANGES_REQUESTED** [review]
train.py had some unresolved merge conflicts. I added suggestions for fixing them, but I recommend running train.py to verify it works as desired.

File: hyperparameter_search.py [hunk 10]
```diff
@@ -297,17 +298,17 @@ def objective(trial, MODEL_NAME, NUM_CLASSES, N_EPOCHS, OPTIMIZER_SEARCH_SPACE,
 
 @click.command()
 @click.option('-m', '--model', required=True, type=click.Choice(AVAILABLE_MODELS, case_sensitive=False), help='Model architechture.')
-@click.option('-e', '--no_of_epochs', type=int, default=20, help='Number of epochs in training loop.')
-@click.option('-t', '--no_of_trials', type=int, default=50, help='Number of hyperparamter search trials in training loop.')
+@click.option('-e', '--no_of_epochs', type=int, show_default=True, default=20, help='Number of epochs in training loop.')
+@click.option('-es', '--early_stopping_counter', type=int, help='Number of consequtive epochs with no improvement in loss until trial is stopped. Default: ~one third of no of epochs.')
```
Comment (jarsba) [inline hyperparameter_search.py:302]: I would limit the maximum patience value to ~10 or so epochs. I would find it very unlikely that with 200 epochs the validation loss would decrease in 66 epochs consecutively.

```suggestion
@click.option('-es', '--early_stopping_counter', type=int, help='Number of consecutive epochs with no improvement in loss until trial is stopped. Default: ~one third of no of epochs or 10, which ever is smaller.')

```

File: hyperparameter_search.py [hunk 11]
```diff
@@ -17,7 +18,7 @@
 from pytorchtools import EarlyStopping
 import warnings
 import logging
-from utils.model_utils import AVAILABLE_MODELS
+from utils.model_utils import AVAILABLE_MODELS, create_model_id_and_timestamp, save_dataset_of_torch_model
 from dataloaders.dataset_stats import get_normalization_mean_std
 from dataloaders.dataset_labels import get_dataset_labels
```
Comment (jarsba) [inline hyperparameter_search.py:24]: Related to the patience suggestion in row 302 and 349

```suggestion
import math
```

File: hyperparameter_search.py [hunk 12]
```diff
@@ -354,24 +358,28 @@ def search_hyperparameters(model, no_of_epochs, no_of_trials, dataset, data_csv,
...
             transforms.ToTensor(),
             transforms.Normalize(mean=mean, std=std),
-            # GaussianNoise(0., 0.1),
         ])
     else:
         data_transform = transforms.Compose([
             transforms.ToPILImage(),
             transforms.Pad(50),
-            transforms.Resize((299, 299)) if model == "inception_v3" else transforms.Resize((256, 256)),
+            transforms.Resize((299, 299)) if MODEL_NAME == "inception_v3" else transforms.Resize((256, 256)),
```
Comment (jarsba) [inline hyperparameter_search.py:366]: I would add get_image_size function to return the supported image size for the model introduced in the commit 206a76fb177a57e5fbbdbe69c021544a27208004 

```suggestion
            transforms.Resize(image_size),
```

File: hyperparameter_search.py [hunk 13]
```diff
@@ -354,24 +358,28 @@ def search_hyperparameters(model, no_of_epochs, no_of_trials, dataset, data_csv,
             transforms.Resize((299, 299)) if MODEL_NAME == "inception_v3" else transforms.Resize((256, 256)),
```
Comment (jarsba) [inline hyperparameter_search.py:358]: I would add get_image_size function to return the supported image size for the model introduced in the commit 206a76fb177a57e5fbbdbe69c021544a27208004 

```suggestion
            transforms.Resize(image_size),
```

File: train.py [hunk 14]
```diff
@@ -124,64 +142,111 @@ def train(model, dataset, data_csv, binary, binary_label, params_file, augmentat
...
+        train_dataset, test_dataset = torch.utils.data.random_split(dataset=master_dataset,
+                                        lengths=[train_size + val_size, test_size],
+                                        generator=torch.Generator().manual_seed(42))
         train_dataset, test_dataset = torch.utils.data.random_split(master_dataset, [train_size, test_size])
-
         train_plant_dataloader = DataLoader(train_dataset, batch_size=BATCH_SIZE_TRAIN, shuffle=True, num_workers=0)
         test_plant_dataloader = DataLoader(test_dataset, batch_size=BATCH_SIZE_TEST, shuffle=False, num_workers=0)
-        # %%
-
-        if torch.cuda.is_available():
```
Comment (jarsba) [inline train.py:150]: Why is the device variable removed? I would guess the "device"-variable is now undefined in line 224 for example.

File: utils/model_utils.py [hunk 15]
```diff
@@ -60,6 +62,13 @@ def get_model_file_name(id: str, model_name: str, timestamp: str) -> str:
 
 	return model_file_name
 
+def get_dataset_file_name(id: str, model_name: str, timestamp: str, prefix: str) -> str:
+	if type(model_name) != str:
+		raise ValueError("Model name must be string")
+
+	model_file_name = f"{id}-{model_name}-{timestamp}-{prefix}.npy"
```
Comment (jarsba) [inline utils/model_utils.py:69]: Small change to variable naming

```suggestion
	dataset_file_name = f"{id}-{model_name}-{timestamp}-{prefix}.npy"
```

File: utils/model_utils.py [hunk 16]
```diff
@@ -60,6 +62,13 @@ def get_model_file_name(id: str, model_name: str, timestamp: str) -> str:
...
+	return model_file_name
```
Comment (jarsba) [inline utils/model_utils.py:70]:
```suggestion
	return dataset_file_name
```

File: hyperparameter_search.py [hunk 17]
```diff
@@ -394,19 +403,29 @@ def search_hyperparameters(model, no_of_epochs, no_of_trials, dataset, data_csv,
         device = torch.device('cuda')
     else:
         device = torch.device('cpu')
+
```
Comment (jarsba) [inline hyperparameter_search.py:406]: I would remove the commented code 407-410

File: hyperparams.yaml [hunk 18]
```diff
@@ -1,4 +1,5 @@
 resnet18:
+  HYPERPARAM_SEARCH_ID: a29yoyfu-resnet18-160522_22:52
```
Comment (jarsba) [inline hyperparams.yaml:2]: I believe these were used to fetch the correct dataset. I would remove these unless they are still useful.

File: hyperparameter_search.py [hunk 19]
```diff
@@ -329,22 +330,25 @@ def search_hyperparameters(model, no_of_epochs, no_of_trials, dataset, data_csv,
...
     N_TRIALS = no_of_trials # Number of trials for hyperparameter optimization
     BATCH_SIZE_TRAIN = 64
     BATCH_SIZE_VALID = 64
-    BATCH_SIZE_TEST = 64
     FLAG_EARLYSTOPPING = True # Set to True to enable early stopping
-    EARLYSTOPPING_PATIENCE = N_EPOCHS//3 # Let early stopping patience (i.e. the number of consequtive epochs with no decrease in training loss) be a third (rounded down) of the number of epochs
+    if early_stopping_counter:
+        EARLYSTOPPING_PATIENCE = early_stopping_counter
+    else:
+        EARLYSTOPPING_PATIENCE = N_EPOCHS//3 # By default early stopping patience (i.e. the number of consequtive epochs with no decrease in training loss) is a third (rounded down) of the number of epochs
```
Comment (jarsba) [inline hyperparameter_search.py:349]:
```suggestion
        EARLYSTOPPING_PATIENCE = math.min(N_EPOCHS//3, 10) # By default early stopping patience (i.e. the number of consecutive epochs with no decrease in training loss) is a third (rounded down) of the number of epochs or 10, which ever is smaller.
```

**Review by jarsba — CHANGES_REQUESTED** [review]
Few small suggestions, Antti's suggestions covered most of my suggestions.

File: train.py (hunk 14 above)
Comment (Jakoviz) [reply train.py:150]: Added just before the training loop on row 213.

File: hyperparameter_search.py (hunk 2 above)
Comment (Jakoviz) [reply hyperparameter_search.py]: Fixed
